src/org/thoughtcrime/securesms/ProfileDocumentsFragment.java

```diff
@@ -172,39 +176,84 @@ public void onMediaLongClicked(DcMsg mediaRecord) {
-          return null;
-        }
-
-      }.execute(mediaRecords.toArray(new DcMsg[mediaRecords.size()]));
+  private void handleDisplayDetails(DcMsg dcMsg) {
+    String info_str = dcContext.getMsgInfo(dcMsg.getId());
+    AlertDialog d = new AlertDialog.Builder(getActivity())
+            .setMessage(info_str)
+            .setPositiveButton(android.R.string.ok, null)
+            .create();
+    d.show();
+    try {
+      //noinspection ConstantConditions
+      Linkify.addLinks((TextView) d.findViewById(android.R.id.message), Linkify.WEB_URLS);
+    } catch(NullPointerException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private void handleDeleteMedia(final Set<DcMsg> messageRecords) {
+    int messagesCount = messageRecords.size();
+
+    new AlertDialog.Builder(getActivity())
+            .setMessage(getActivity().getResources().getQuantityString(R.plurals.ask_delete_messages, messagesCount, messagesCount))
+            .setCancelable(true)
+            .setPositiveButton(R.string.delete, (dialog, which) -> {
+                int[] ids = DcMsg.msgSetToIds(messageRecords);
+                dcContext.deleteMsgs(ids);
+                actionMode.finish();
+            })
+            .setNegativeButton(android.R.string.cancel, null)
+            .show();
+  }
+
+  private void handleShowInChat(final DcMsg dcMsg) {
+    Intent intent = new Intent(getContext(), ConversationActivity.class);
+    intent.putExtra(ConversationActivity.CHAT_ID_EXTRA, dcMsg.getChatId());
+    intent.putExtra(ConversationActivity.STARTING_POSITION_EXTRA, DcMsg.getMessagePosition(dcMsg, dcContext));
+    startActivity(intent);
+  }
+
+  private void handleSaveAttachment(final DcMsg message) {
+    SaveAttachmentTask.showWarningDialog(getContext(), (dialogInterface, i) -> {
+        Permissions.with(getActivity())
+                .request(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE)
+                .ifNecessary()
+                .withPermanentDenialDialog(getString(R.string.perm_explain_access_to_storage_denied))
+                .onAllGranted(() -> {
+                    SaveAttachmentTask saveTask = new SaveAttachmentTask(getContext());
+                    SaveAttachmentTask.Attachment attachment = new SaveAttachmentTask.Attachment(
+                            Uri.fromFile(message.getFileAsFile()), message.getFilemime(), message.getDateReceived(), message.getFilename());
+                    saveTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, attachment);
+                    actionMode.finish();
+                })
+                .execute();
     });
-    builder.setNegativeButton(android.R.string.cancel, null);
-    builder.show();
+  }
+
+  private DcMsg getSelectedMessageRecord() {
+    Set<DcMsg> messageRecords = getListAdapter().getSelectedMedia();
+
+    if (messageRecords.size() == 1) return messageRecords.iterator().next();
+    else                            throw new AssertionError();
```

Comment:
Does multi-selection work with this PR? (I didn't build it, but here only single-selection seems to be allowed)

Reply:
yes multiselection works, that only means that multiselection is not allowed for "share" and "export attachments" that exception is never thrown because menu items are hidden when multiple are selected, it is just an extra precaution that the items are not left visible by accident

Reply:
it means "give me one element, I expect only one element to be selected", btw, I just copy+pasted `getSelectedMessageRecord()` from `ConversationFragment.java` :D

Reply:
it makes sense to group all that utilities somehow in one common class or so

Reply:
This method should probably go into MessageSelectorFragment, then?

Reply:
@Hocuri to move `getSelectedMessageRecord()` to MessageSelectorFragment would then need to refactor the three adapters under a common class/interface that has `getSelectedMessage()`, that is if I add it with implementation, I can add it as abstract but that will not help in the code duplication

Reply:
solved